### PR TITLE
Don't set `LD_LIBRARY_PATH` on package load

### LINKED ISSUE #3955: Setting LD_LIBRARY_PATH to be "helpful" considered harmful
(with apologies to Djikstra and everyone else who's recycled that title meme)

**TL;DR**: the library that one of my spack binaries uses depends on what other spack packages I've `module load`-ed.  **YIKES**.  See also #3926.

I was trying to understand why @JusticeForMikeBrown was having trouble building bowtie2 (see #3950) when I've built it successfully with `gcc@4.x.y`.

His problem with `gcc@4.x.y` was zlib related; I checked the package and noticed that it doesn't have a dependency on zlib.  Perhaps it should, I thought.  Wonder what zlib my "production" copy was linked against?

```
$ ldd bowtie2-align-l  | grep libz
	libz.so.1 => /blah/spack/v0.0.8/opt/spack/linux-centos7-x86_64/gcc-5.4.0/zlib-1.2.11-ec535e2ikkpl7hd4y454t3yydjqorja6/lib/libz.so.1 (0x00002aaaaaf32000)
```

That surprised me, because there's no zlib dependency in the package.

Sure enough, it's because I have something else `module load`-ed that has the side effect of adding zlib's directory to `LD_LIBRARY_PATH`.

```
$ (unset LD_LIBRARY_PATH; ldd bowtie2-align-l) | grep libz
	libz.so.1 => /lib64/libz.so.1 (0x00002aaaaaf2f000)
```

My "newer" version of CentOS has a `/lib64/libz.so.1` that includes gzbuffer (`nm` didn't help, library's stripped...):

```
$ strings /lib64/libz.so.1 | grep buffer
gzbuffer
buffer error
```

so it (probably) works for me either way.

But imagine if there were two versions of a library (perhaps something mathematical) that give different results.  Now you have a program giving different results depending on what other Spack applications are also loaded.  

**THAT** would be fun to track down (assuming you even noticed...).

W.R.T. the main problem, bowtie2 should probably have a dependency on a new-ish version of zlib, but stuff like this is why LD_LIBRARY_PATH is a slippery tool to reach for.

I'll argue that this kind of unpredictability is a bigger negative than being helpful and always setting `LD_LIBRARY_PATH`.  This comment in the docs isn't actually correct:

> Spack avoids library misconfiguration by using RPATH to link dependencies. When a user links a library or runs a program, it is tied to the dependencies it was built with, so there is no need to manipulate LD_LIBRARY_PATH at runtime.
>
> [clipped from here](http://spack.readthedocs.io/en/latest/features.html?highlight=RPATH#packages-can-peacefully-coexist)

What would happen if `LD_LIBRARY_PATH` became *opt-in*, packages that need it specify it in their package definitions?

Looking at [the list of cases where RPATH support doesn't work](http://spack.readthedocs.io/en/latest/workflows.html?highlight=LD_LIBRARY_PATH#transitive-dependencies), it seems like 1) is not relevant (I think it's referring to `PERL5LIB`, etc...) and 3) are simply bugs.  That leaves 2), python extensions.  Is `RPATH` unworkable there or just not yet working?

### LINKED ISSUE #31731: symbol lookup error: /lib64/libk5crypto.so.3: undefined symbol: EVP_KDF_ctrl, version OPENSSL_1_1_1b
### Steps to reproduce

For me, this happens whenever I launch `emacs` after loading a Spack-built OpenSSL.
```console
$ spack install openssl@1.1.1q
$ spack load openssl@1.1.1q
```

### Error message

```console
$ emacs
emacs: symbol lookup error: /lib64/libk5crypto.so.3: undefined symbol: EVP_KDF_ctrl, version OPENSSL_1_1_1b
```

### Information on your system

```console
$ spack debug report
* **Spack:** 0.19.0.dev0 (b065d69136023693ace8e4d35d9fdad817d63d93)
* **Python:** 3.6.8
* **Platform:** linux-rhel8-zen2
* **Concretizer:** clingo
```
The system-installed version of OpenSSL is 1.1.1k

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by haampie**
For me this has been mostly just annoying, cause if I do 
`spack env activate ...` and it sets LD_LIBRARY_PATH, system executables that don't
use rpath will now consider spack-built libs first, and are sometimes
incompatible.

And executables built by Spack already use rpath anyways, so they know
where there libs are.

The only case is maybe when libs are dlopen'ed by say Python, but even
then, I think the libs are generally located because of Python's own
rpaths set by Spack.

Closes #3955 #31731

### CODE REVIEW

File: etc/spack/defaults/cray/modules.yaml [hunk 1]
```diff
@@ -13,9 +13,4 @@
 # Per-user settings (overrides default and site settings):
 #   ~/.spack/modules.yaml
 # -------------------------------------------------------------------------
-modules:
-  prefix_inspections:
-    lib:
-      - LD_LIBRARY_PATH
-    lib64:
-      - LD_LIBRARY_PATH
+modules: {}
```
Comment (adamjstewart) [inline etc/spack/defaults/cray/modules.yaml:16]: If these files are no longer being used they can just be removed

**Review by adamjstewart — COMMENTED** [review]
This PR will break most Python/SCons packages with external dependencies. Python does not properly support RPATHs.

File: etc/spack/defaults/cray/modules.yaml (hunk 1 above)
Comment (haampie) [reply etc/spack/defaults/cray/modules.yaml:16]: That made the tests fail :( didn't check why, but then I thought it wouldn't hurt as a blueprint

**Comment by haampie** [comment]
Can't PythonPackage set LD_LIBRARY_PATH for link type deps that are meant to be dlopen'ed? Just like PYTHONPATH

**Comment by adamjstewart** [comment]
It can, although that will make it even harder to get rid of `setup_dependent_*_environment`.

File: etc/spack/defaults/darwin/modules.yaml [hunk 2]
```diff
@@ -13,9 +13,4 @@
 # Per-user settings (overrides default and site settings):
 #   ~/.spack/modules.yaml
 # -------------------------------------------------------------------------
-modules:
-  prefix_inspections:
-    lib:
-      - DYLD_FALLBACK_LIBRARY_PATH
-    lib64:
-      - DYLD_FALLBACK_LIBRARY_PATH
```
Comment (adamjstewart) [inline etc/spack/defaults/darwin/modules.yaml]: I understand your concern with `LD_LIBRARY_PATH`, but this issue does not arise with `DYLD_FALLBACK_LIBRARY_PATH`. On macOS, there are two env vars that control library selection at load-time: `DYLD_LIBRARY_PATH` and `DYLD_FALLBACK_LIBRARY_PATH`. The former acts just like `LD_LIBRARY_PATH` and can break system exes that aren't RPATHed. The latter acts more like RUNPATH and provides libraries to use only when the libraries aren't found in normal search locations. So I think it's safe to keep these here.

**Comment by haampie** [comment]
> This PR will break most Python packages with external dependencies. Python does not properly support RPATHs.

Is this really true @adamjstewart? Don't python packages usually call C functions through a generated wrapper library that does have rpaths already?

For instance when running `./bin/python` directly without activating an environment, and running `import numpy`, I get:

```
$ cat /proc/451559/maps | grep openblas
7f7218ad1000-7f7218b40000 r--p 00000000 103:04 40283657                  /home/harmen/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-10.3.0/openblas-0.3.19-ixyrqkwedhttcxokzzczho4zefy7b3ji/lib/libopenblas_zen-r0.3.19.so
```

so it manages to load libraries without LD_LIBRARY_PATH just fine.

And can you explain what you mean with SCons given that LD_LIBRARY_PATH is runtime.

File: etc/spack/defaults/darwin/modules.yaml (hunk 2 above)
Comment (haampie) [reply etc/spack/defaults/darwin/modules.yaml]: Yeah, reverted that, it's just about linux now.

**Comment by salotz-sitx** [comment]
When I load an environment and `LD_LIBRARY_PATH` is set this sometimes conflicts with the distro installed packages that depend on dynamic linking via that variable. This can cause segfaults to occur depending on the versions. I haven't experienced a segfault myself (Ubuntu 18.04) but a coworker has with Ubuntu 21.04 with `nano`.

E.g.:

```
--> ldd /usr/bin/python3
	linux-vdso.so.1 (0x00007ffc9603d000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fce3ef58000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fce3ed39000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fce3eb35000)
	libutil.so.1 => /lib/x86_64-linux-gnu/libutil.so.1 (0x00007fce3e932000)
	libexpat.so.1 => /home/salotz/scratch/mdutil/.spack-env/view/lib/libexpat.so.1 (0x00007fce3e6fe000)
	libz.so.1 => /home/salotz/scratch/mdutil/.spack-env/view/lib/libz.so.1 (0x00007fce3e4e7000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fce3e149000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fce3f349000)
	libbsd.so.0 => /home/salotz/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/libbsd-0.11.3-rkyd4ndtgypgsbhazhd4dwbflwskzf36/lib/libbsd.so.0 (0x00007fce3df34000)
	libmd.so.0 => /home/salotz/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/libmd-1.0.3-fwt4srl6qp7veimuan65dc23zzobjkpr/lib/libmd.so.0 (0x00007fce3dd28000)
```

**Comment by adamjstewart** [comment]
> Is this really true @adamjstewart? Don't python packages usually call C functions through a generated wrapper library that does have rpaths already?

Most Python libraries use `ctypes.util.find_library`, which searches typical system locations and environment variables only. Packages like numpy which hardcode the library location in `setup.cfg` are the exception, not the rule. If you want to reproduce this, try `py-shapely`, that's the first example that comes to mind. The same is true for most geospatial Python libs.

> Can't PythonPackage set LD_LIBRARY_PATH for link type deps that are meant to be dlopen'ed? Just like PYTHONPATH

If we do push through with this PR, we'll have to do this if we don't want to break hundreds of packages.

File: lib/spack/spack/user_environment.py [hunk 3]
```diff
@@ -32,21 +32,18 @@ def prefix_inspections(platform):
 
     inspections = {
         'bin': ['PATH'],
-        'lib': ['LD_LIBRARY_PATH', 'LIBRARY_PATH'],
-        'lib64': ['LD_LIBRARY_PATH', 'LIBRARY_PATH'],
         'man': ['MANPATH'],
         'share/man': ['MANPATH'],
         'share/aclocal': ['ACLOCAL_PATH'],
-        'include': ['CPATH'],
```
Comment (ChristianTackeGSI) [inline lib/spack/spack/user_environment.py]: Why do you remove `include` / `CPATH` as well in this PR?

I could be discussed, whether people wanting to build packages against other packages, they should know what they're doing, etc.

But really, this is a completely different discussion. IMHO it should be in its own PR/Issue?

Comment (haampie) [reply lib/spack/spack/user_environment.py]: This is the fallback when etc/defaults/modules.yaml is missing for some reason (which is part of spack sources), and default config doesn't have CPATH. This was just an oversight. But fair, it's technically not part of the pr.

**Comment by adamjstewart** [comment]
I know I've been the main person opposing this for years, but I'm starting to warm up to this idea. I think the packages that require LD_LIBRARY_PATH are few enough that we can special case them. Now might be a good time to merge this and see what breaks since we just had a release.

**Comment by haampie** [comment]
So, should we merge it?

**Comment by adamjstewart** [comment]
I'm fine with that but would like to get the opinions of a couple more core devs first.

**Comment by tgamblin** [comment]
I am strongly in favor of this change -- the question is what to do for people who relied on this stuff.  Basically I think Spack should work and loading an environment should not affect your compiler or `ld.so`.  So I agree with removing:

* `LD_LIBRARY_PATH`
* `LIBRARY_PATH`
* `CPATH`

The last two AFAIK don't work on all compilers anyway, so it's at best inconsistent to set them.

@haampie: could you modify this, though, so that we set these?

* `SPACK_LIBRARY_PATH`
* `SPACK_INCLUDE_PATH`

to the prior values of `LD_LIBRARY_PATH`/`LIBRARY_PATH` and `CPATH`, respectively?  And document it?

That way someone who loads an env will have something they can use to set the variables they need, but we won't be borking the user environment by default.

**Comment by tgamblin** [comment]
I added this here since it's a fairly impactful change: https://github.com/spack/spack/discussions/30634

**Comment by haampie** [comment]
Note that `CPATH` and `LIBRARY_PATH` were already removed long ago. To hit the code path in `prefix_inspections` where you get `CPATH` and `LIBRARY_PATH`, you need to delete all `modules.yaml` config files including `$spack/etc/defaults/modules.yaml`, which seems unlikely. This PR just fixes some oversights wrt CPATH/LIBRARY_PATH.

So this PR is really about `LD_LIBRARY_PATH` only.

> could you modify this, though, so that we set these?

Do you mean to store `<prefix>/lib(64)?` in `SPACK_LD_LIBRARY_PATH` so users can do `LD_LIBRARY_PATH=$SPACK_LD_LIBRARY_PATH` if they want the previous behavior? I would say users can just change the config files, right? E.g.
```
spack config add modules:prefix_inspections:lib64:[LD_LIBRARY_PATH]
spack config add modules:prefix_inspections:lib:[LD_LIBRARY_PATH]
```

Also, can we run a pipeline from scratch using this PR? It affects the builds too (note that `modules:prefix_inspections` is also used to set up the build environment).

**Comment by tgamblin** [comment]
> Do you mean to store <prefix>/lib(64)? in `SPACK_LD_LIBRARY_PATH` so users can do `LD_LIBRARY_PATH=$SPACK_LD_LIBRARY_PATH` if they want the previous behavior?

Yes.

```
spack config add modules:prefix_inspections:lib64:[LD_LIBRARY_PATH]
spack config add modules:prefix_inspections:lib:[LD_LIBRARY_PATH]
```

They could do this, but IMO it's a little clunky, and having the env vars allows you do do more -- e.g., you may want to pass the list of transitive library search directories to CMake or use it on your link line or something.  I'm basically saying that we should have a convention for that and throw it in a harmless env var (as opposed to a harmful one like now).

Users will be familiar with this from module schemes like OpenHPC, where they have a similar policy. I think it's a good default.

**Comment by adamjstewart** [comment]
I personally agree with @haampie, if people want to set additional env vars in module files, Spack already has a way to do that, we don't need to set `SPACK_*` env vars for them to use instead.

**Comment by tgamblin** [comment]
> I personally agree with @haampie, if people want to set additional env vars in module files, Spack already has a way to do that, we don't need to set SPACK_* env vars for them to use instead.

Ok fair enough -- I'll relent.  They can add those settings via `spack config`.  Can that please be documented here, though?

**Comment by adamjstewart** [comment]
Worth documenting in the release notes too since it's a backwards incompatible change.

**Comment by tgamblin** [comment]
I added it to the high impact list over in #30634 and that always also goes in the release notes.

@haampie: there will one day very soon be a `spackbot rebuild everything`, but for now to build the world here, you have to do something like [e4a9dc4](https://github.com/spack/spack/pull/31752/commits/e4a9dc4d3e5f11426eb209128aba42e32bbb3e25) in #31752.  Basically, make another PR that cherry-picks that commit and push it to trigger a full rebuild.

**Comment by haampie** [comment]
I don't see issues in #32019 apart from mpfr not building because it's known to be broken

**Review by tgamblin — APPROVED** [review]

**Comment by h-murai** [comment]
How about the case where you load a library by Spack and then link your objects with the library?
```
$ spack load fftw
$ gcc foo.c -lfftw3
```
Does it work after this change?

**Comment by jbigot** [comment]
> How about the case where you load a library by Spack and then link your objects with the library?
> 
> ```
> $ spack load fftw
> $ gcc foo.c -lfftw3
> ```
> 
> Does it work after this change?

This is my exact use-case (although with en environment) we distribute a spack env that people wanting to hack on our just have to load to start working. I recently noticed `LD_LIBRARY_PATH` is not set by the env anymore. With cmake, deps are still found ok, but as soon as you try to run the compiled code, it doesn't work anymore.

**Comment by jbigot** [comment]
could a flag be provided to explicitly ask for `LD_LIBRARY_PATH` to be set?

**Comment by adamjstewart** [comment]
The set of environment variables added by Spack is completely customizable. Although `LD_LIBRARY_PATH` is no longer added by default, you can add it to your Spack installation like so:
```console
$ spack config add modules:prefix_inspections:lib64:[LD_LIBRARY_PATH]
$ spack config add modules:prefix_inspections:lib:[LD_LIBRARY_PATH]
```
After this, `spack load ...` will work identically to the way that it used to. See https://spack.readthedocs.io/en/latest/module_file_support.html#write-a-configuration-file for examples of what this will look like, the above command just writes this file for you. Note that you may need to specify that this is added to your environment, or to a global scope so that it affects all users, not just you.

Does that address your concerns?

**Comment by jbigot** [comment]
thank you for the pointer. Indeed, this should do the trick!

**Comment by fspiga** [comment]
Today I discovered while I was doing a fresh build, I discovered generated TCL modules do not export anymore `LD_LIBRARY_PATH`.

Spack has been an extremely useful tool to install packages, self-build modules, expose to users these modules that can be used by users using `module load` in the old fashion way. Many end users do not really need to know spack exists or learning spack commands. 

**Please** documents very visibly the workaaround
```
$ spack config add modules:prefix_inspections:lib64:[LD_LIBRARY_PATH]
$ spack config add modules:prefix_inspections:lib:[LD_LIBRARY_PATH]
```
because a lot of your users and sysadmins will likely need it.

**Comment by haampie** [comment]
Feel free to make a PR
